third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -223,22 +290,49 @@ static SmallVector<Value> Fp8E5M2_to_Fp16(Location loc,
           extract_element(f16_ty, fp16x2Vec1, i32_val(0)),
           extract_element(f16_ty, fp16x2Vec1, i32_val(1))};
 }
-#else
-static const std::string Fp8E5M2_to_Fp16(bool hasNativeFP) {
-  std::string ret;
-  if (!hasNativeFP) {
-    ret = "{                           \n"
-          "prmt.b32 $0, 0, $2, 0x5140; \n\t"
-          "prmt.b32 $1, 0, $2, 0x7362; \n\t"
-          "}";
-  } else {
-    ret = "cvt.rn.f16x2.e5m2x2 $0, $1; \n\t";
+
+static Value convertBf16ToFp32(Location loc,
+                               ConversionPatternRewriter &rewriter,
+                               const Value &v) {
+  auto as_int16 = bitcast(v, i16_ty);
+  auto as_int32 = zext(i32_ty, as_int16);
+  auto shifted = shl(i32_ty, as_int32, i32_val(16));
+  return (bitcast(shifted, f32_ty));
+}
+
+static Value convertFp32ToBf16(Location loc,
+                               ConversionPatternRewriter &rewriter,
+                               const Value &v, const RoundingMode rounding) {
+  if (rounding == RoundingMode::RTNE) {
+    auto as_uint32 = bitcast(v, i32_ty);
+    auto check_exponent =
+        and_(i32_ty, xor_(i32_ty, as_uint32, i32_val(0xffffffff)),
+             i32_val(0x7f800000));
+    auto exponent_not_all1s = icmp_ne(check_exponent, i32_val(0));
+    auto exponent_all1s = icmp_eq(check_exponent, i32_val(0));
+    auto rounded =
+        add(i32_ty, i32_val(0x7fff),
+            and_(i32_ty, lshr(i32_ty, as_uint32, i32_val(16)), i32_val(1)));
+    rounded = add(i32_ty, rounded, as_uint32);
+    auto res = select(exponent_not_all1s, rounded, as_uint32);
+
+    auto preserve_nan =
+        and_(i1_ty, exponent_all1s,
+             icmp_ne(and_(i32_ty, as_uint32, i32_val(0xffff)), i32_val(0)));
+    auto nan = or_(i32_ty, as_uint32, i32_val(0x10000));
+    res = select(preserve_nan, nan, res);
+
+    auto shifted = lshr(i32_ty, res, i32_val(16));
+    auto truncated = trunc(i16_ty, shifted);
+    return truncated;
+  } else { // if (rounding == RoundingMode::RTZ)
+    auto as_int32 = bitcast(v, i32_ty);
```

Comment:
You can move this case to the beginning and do early return. So for the majority of the code in the above we can drop one level of indent.

Reply:
Done.